core/server/proxy/src/main/java/alluxio/proxy/s3/signature/AwsSignatureProcessor.java

```diff
@@ -0,0 +1,229 @@
+    } else {
+      return new SignatureInfo(
+              SignatureInfo.Version.NONE,
+              "", "", "", "", "", "", "", false
+      );
+    }
+  }
+
+  /**
+   * Convert SignatureInfo to AwsAuthInfo.
+   * @return AwsAuthInfo
+   * @throws S3Exception
+   */
+  public AwsAuthInfo getAuthInfo() throws S3Exception {
+    try {
+      SignatureInfo signatureInfo = parseSignature();
+      String stringToSign = "";
+      if (signatureInfo.getVersion() == SignatureInfo.Version.V4) {
+        stringToSign =
+                StringToSignProducer.createSignatureBase(signatureInfo, mContext);
+      }
+      String awsAccessId = signatureInfo.getAwsAccessId();
+      // ONLY validate aws access id when needed.
+      if (StringUtils.isEmpty(awsAccessId)) {
+        LOG.debug("Malformed s3 header. awsAccessID is empty");
+        throw new S3Exception("awsAccessID is empty", S3ErrorCode.ACCESS_DENIED_ERROR);
+      }
+
+      return new AwsAuthInfo(awsAccessId,
+              stringToSign,
+              signatureInfo.getSignature()
+              );
+    } catch (S3Exception ex) {
+      LOG.debug("Error during signature parsing: ", ex);
+      throw ex;
+    } catch (Exception e) {
+      // For any other critical errors during object creation throw Internal
+      // error.
+      LOG.debug("Error during signature parsing: ", e);
+      throw new S3Exception(e, "Context is invalid", S3ErrorCode.INTERNAL_ERROR);
+    }
+  }
+
+  /**
+   * A simple map which forces lower case key usage.
+   */
+  public static class LowerCaseKeyStringMap implements Map<String, String> {
+    private Map<String, String> mDelegate;
+
+    /**
+     * Constructs a new {@link LowerCaseKeyStringMap}.
+     */
+    public LowerCaseKeyStringMap() {
+      mDelegate = new HashMap<>();
+    }
+
+    /**
+     * Constructs a new {@link LowerCaseKeyStringMap} from headers.
+     * @param rawHeaders
+     * @return a LowerCaseKeyStringMap instance
+     */
+    public static LowerCaseKeyStringMap fromHeaderMap(
+        MultivaluedMap<String, String> rawHeaders) {
+
+      //header map is MUTABLE. It's better to save it here. (with lower case
+      // keys!!!)
+      final LowerCaseKeyStringMap headers = new LowerCaseKeyStringMap();
+
+      for (Entry<String, List<String>> headerEntry : rawHeaders.entrySet()) {
+        if (0 < headerEntry.getValue().size()) {
+          String headerKey = headerEntry.getKey();
+          if (headers.containsKey(headerKey)) {
+            //multiple headers from the same type are combined
+            headers.put(headerKey,
+                    String.format("%s,%s", headers.get(headerKey), headerEntry.getValue().get(0)));
+          } else {
+            headers.put(headerKey, headerEntry.getValue().get(0));
+          }
+        }
+      }
+
+      if (LOG.isTraceEnabled()) {
+        headers.keySet().forEach(k -> LOG.trace("Header:{},value:{}", k,
+             headers.get(k)));
+      }
+      return headers;
+    }
```

Comment:
From what I can tell, you are using this class to perform two tasks:
1. Allow "ignore case" behaviour for writing/looking up request Headers

Why is forcing everything to lowercase necessary? From what I understand the header keys are well-defined in S3 documentation and you can directly use their values as-is. You can put them in `SignerConstants` like you have done for the query parameters.

2. Handling duplicate headers by appending their contents together

Can you provide an example why this is necessary? From what I understand these headers/query parameters should not be provided more than once. If they are that is an invalid request on the client's part and we can either explicitly check for that and fail the request (preferred) or we can silently ignore it and have unknown behavior. I don't think we need to explicitly handle it ourselves like is being done here.

Reply:
1.SignedHeaders are all lower case and need to get these value from the HeaderMap and compute signature need lower case for headers.
     
2. Removed duplicate headers.